src/payment/strategies/masterpass/masterpass-payment-strategy.ts

```diff
@@ -0,0 +1,121 @@
+
+        return super.deinitialize(options);
+    }
+
+    execute(payload: OrderRequestBody, options?: PaymentRequestOptions): Promise<InternalCheckoutSelectors> {
+        const order = { useStoreCredit: payload.useStoreCredit };
+
+        if (!this._paymentMethod) {
+            throw new MissingDataError(MissingDataErrorType.MissingPaymentMethod);
+        }
+
+        const methodId = this._paymentMethod.initializationData.gateway;
+        const paymentData = this._paymentMethod.initializationData.paymentData;
+
+        if (!methodId) {
+            const msg = 'Unable to proceed because "payload.payment.methodId" argument is not provided.';
+            throw new InvalidArgumentError(msg);
+        }
+
+        if (!paymentData) {
+            const msg = 'Unable to proceed because "payload.payment.paymentData" argument is not provided.';
+            throw new InvalidArgumentError(msg);
+        }
+
+        return this._store.dispatch(this._orderActionCreator.submitOrder(order, options))
+            .then(() => this._store.dispatch(this._paymentMethodActionCreator.loadPaymentMethod(methodId)))
+            .then(() => this._store.dispatch(this._paymentActionCreator.submitPayment({ methodId, paymentData })));
+    }
+
+    private _createMasterpassPayload(paymentMethod: PaymentMethod, checkout: Checkout, storeConfig: StoreConfig): MasterpassCheckoutOptions {
+        return {
+            checkoutId: paymentMethod.initializationData.checkoutId,
+            allowedCardTypes: paymentMethod.initializationData.allowedCardTypes,
+            amount: checkout.subtotal.toFixed(2),
+            currency: storeConfig.currency.code,
+            cartId: checkout.cart.id,
+            suppressShippingAddress: true,
+        };
+    }
+
+    private _hasPaymentInfo(): boolean {
+        return this._paymentMethod
+            && this._paymentMethod.initializationData
+            && this._paymentMethod.initializationData.paymentData
+            && this._paymentMethod.initializationData.paymentData.nonce;
+    }
+
+    private _masterpassClientSetup(): Promise<MasterpassCheckoutCallback> {
+        const state = this._store.getState();
+        const checkout = state.checkout.getCheckout();
+        const storeConfig = state.config.getStoreConfig();
+
+        if (!checkout) {
+            throw new MissingDataError(MissingDataErrorType.MissingCheckout);
+        }
+
+        if (!storeConfig) {
+            throw new MissingDataError(MissingDataErrorType.MissingCheckoutConfig);
+        }
+
+        if (!(this._paymentMethod && this._paymentMethod.initializationData)) {
+            throw new MissingDataError(MissingDataErrorType.MissingPaymentMethod);
+        }
+
+        const payload = this._createMasterpassPayload(this._paymentMethod, checkout, storeConfig);
+
+        return this._masterpassScriptLoader.load(this._paymentMethod.config.testMode)
+            .then(masterpassClient => () => masterpassClient.checkout(payload));
```

Comment:
Actually, reading this code, why are we returning a promise that we then call? Can't we just call the thing directly?

```js
return this._masterpassScriptLoader.load(this._paymentMethod.config.testMode)
	.then(masterpassClient => masterpassClient.checkout(payload));
```

Reply:
What I'm returning in the resolved promise is a function which has the masterpass client configured with the checkout method ready to be called, this is useful for delay the execution, i.e if we pass it to an event handler.

However if we aren't going to pass it to anyone and the redirect will be executed from the initialize method, this can be called directly as you suggested.